gateway-js/src/__tests__/integration/scope.test.ts

```diff
@@ -0,0 +1,116 @@
+import gql from 'graphql-tag';
+import { execute } from '../execution-utils';
+
+import {
+  astSerializer,
+  queryPlanSerializer,
+} from 'apollo-federation-integration-testsuite';
+
+expect.addSnapshotSerializer(astSerializer);
+expect.addSnapshotSerializer(queryPlanSerializer);
+
+describe('scope', () => {
+  it("doesn't wrap inline fragments with the supertype when @include is used", async () => {
+    const query = `#graphql
+    query GetProducts {
+      topProducts {
+        name
+        ... on Shoe @include(if: true) {
+          rating
+        }
+        ... on Car {
+          rating
+        }
+      }
+    }
+  `;
+
+    const { queryPlan, errors } = await execute({ query }, [
+      {
+        name: 'products',
+        typeDefs: gql`
+          extend type Query {
+            topProducts: [Product]
+          }
+
+          interface Product {
+            name: String
+          }
+
+          type Shoe implements Product @key(fields: "upc") {
+            upc: String
+            name: String
+          }
+
+          type Car implements Product @key(fields: "upc") {
+            upc: String
+            name: String
+          }
+        `,
+      },
+      {
+        name: 'reviews',
+        typeDefs: gql`
+          extend type Shoe @key(fields: "upc") {
+            upc: String @external
+            rating: Int
+          }
+
+          extend type Car @key(fields: "upc") {
+            upc: String @external
+            rating: Int
+          }
+        `,
+      },
+    ]);
+
+    expect(errors).toBeUndefined();
+    expect(queryPlan).toMatchInlineSnapshot(`
+      QueryPlan {
+        Sequence {
+          Fetch(service: "products") {
+            {
+              topProducts {
+                __typename
+                ... on Car {
+                  name
+                  __typename
+                  upc
+                }
+                ... on Shoe {
+                  name
+                }
+                ... on Shoe @include(if: true) {
```

Comment:
The query plan in the original failing test was the result from running under 0.25.1. I agree this better reflects the intent, as there is no need to query for `upc` unless the include condition is met. 

It may be worth double checking if it also does the right thing under different circumstances though. I'm thinking of:
- selecting `rating` both under the include condition and without it, where the order of the inline fragments may also matter (because some of the code for [propagating directives is flawed](https://github.com/apollographql/federation/commit/31a7ba605b4931e5bb62d766637ffe005bd2af95)).
- selecting a different extension field of `Shoe` without an include condition, to ensure `upc` is fetched unconditionally in that case.

Reply:
Added a few tests and afaict, it seems to work properly (there is some ugliness in the "argument" part of the dependent fetch related to your other comment but I think that's harmless). I'm not 100% sure I captured what you had in mind for the 1st bullet though. I'm not sure if you mean `rating` to in nested fragments or not and why order would matter. Let me know if there is another case you had in mind that isn't captured.

Reply:
@pcmanus This probably doesn't matter specifically for this, but what I was getting at with the ordering was an issue that I now realize may actually have been fixed by your scoping changes. Originally, we only took directives on the first field nested under a particular type condition into account. So here for example, both `rating` and `reviewsCount` would be considered to fall under `@include(if: true)`:

```
... on Shoe @include(if: true) {
  rating
}

... on Shoe {
  rating
  reviewCount
}
```

And if you changed the order, the condition would be lost completely:
```
... on Shoe {
  rating
  reviewCount
}

... on Shoe @include(if: true) {
  rating
}
```

Reply:
I see. Added a quick test for that exact case (which does pass).